File: typescript/packages/node/src/core/email/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { IndexCacheStore, PathSpec } from '@struktoai/mirage-core'
import { IndexEntry, PathSpec as PathSpecCtor, stripSlash } from '@struktoai/mirage-core'
import type { EmailAccessor } from '../../accessor/email.ts'
import { fetchHeaders, listMessageUids } from './_client.ts'
import { listFolders } from './folders.ts'
import type { ParsedAttachment } from './_parse.ts'

const TITLE_MAX = 80
const UNSAFE = /[^\w\s\-.]/g
const MULTI_UNDERSCORE = /_+/g

function sanitize(text: string): string {
  if (text.trim() === '') return 'No_Subject'
  let cleaned = text.replace(UNSAFE, '_').replace(/ /g, '_')
  cleaned = cleaned.replace(MULTI_UNDERSCORE, '_').replace(/^_+|_+$/g, '')
  if (cleaned.length > TITLE_MAX) cleaned = `${cleaned.slice(0, TITLE_MAX - 3)}...`
  return cleaned
}

function msgFilename(subject: string, uid: string): string {
  return `${sanitize(subject)}__${uid}.email.json`
}

function dateFromHeader(dateStr: string): string {
  if (dateStr === '') return '1970-01-01'
  const d = new Date(dateStr)
  if (Number.isNaN(d.getTime())) return '1970-01-01'
  const yyyy = d.getUTCFullYear().toString().padStart(4, '0')
  const mm = (d.getUTCMonth() + 1).toString().padStart(2, '0')
  const dd = d.getUTCDate().toString().padStart(2, '0')
  return `${yyyy}-${mm}-${dd}`
}

function enoent(p: string): Error {
  const e = new Error(`ENOENT: ${p}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

export async function readdir(
  accessor: EmailAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  const prefix = path.prefix
  const raw = path.pattern !== null ? path.directory : path.original
  let p = raw
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  const virtualKey = key !== '' ? `${prefix}/${key}` : prefix !== '' ? prefix : '/'
  const parts = key === '' ? [] : key.split('/')
  const depth = parts.length

  if (depth === 0) {
    if (index !== undefined) {
      const cached = await index.listDir(virtualKey)
      if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    }
    const folders = await listFolders(accessor)
    const entries: [string, IndexEntry][] = []
    for (const folderName of folders) {
      const entry = new IndexEntry({
        id: folderName,
        name: folderName,
        resourceType: 'email/folder',
        vfsName: folderName,
      })
      entries.push([folderName, entry])
    }
    if (index !== undefined) await index.setDir(virtualKey, entries)
    return entries.map(([name]) => `${prefix}/${name}`)
  }

  if (depth === 1) {
    const folderName = parts[0] ?? ''
    if (index !== undefined) {
      const cached = await index.listDir(virtualKey)
      if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    }
    if (index === undefined) throw enoent(path.original)
    const maxMessages = accessor.config.maxMessages
    const uids = await listMessageUids(accessor, folderName, 'ALL', maxMessages)
    const headersList = await fetchHeaders(accessor, folderName, uids)
    const dateGroups = new Map<string, typeof headersList>()
    for (const hdr of headersList) {
      const dateStr = dateFromHeader(hdr.date)
      let bucket = dateGroups.get(dateStr)
      if (bucket === undefined) {
        bucket = []
        dateGroups.set(dateStr, bucket)
      }
      bucket.push(hdr)
    }
    const sortedDates = [...dateGroups.keys()].sort().reverse()
    const dateEntries: [string, IndexEntry][] = []
    for (const dateStr of sortedDates) {
      const dateEntry = new IndexEntry({
        id: dateStr,
        name: dateStr,
        resourceType: 'email/date',
        vfsName: dateStr,
      })
      dateEntries.push([dateStr, dateEntry])
      const msgEntries: [string, IndexEntry][] = []
      for (const hdr of dateGroups.get(dateStr) ?? []) {
        const uid = hdr.uid
        const subject = hdr.subject || 'No Subject'
        const filename = msgFilename(subject, uid)
        const msgEntry = new IndexEntry({
          id: uid,
          name: subject,
          resourceType: 'email/message',
          vfsName: filename,
        })
        msgEntries.push([filename, msgEntry])
        const attachments: ParsedAttachment[] = hdr.attachments
        if (attachments.length > 0) {
          const attDirName = filename.replace('.email.json', '')
          const attDirEntry = new IndexEntry({
            id: uid,
            name: attDirName,
            resourceType: 'email/attachment_dir',
            vfsName: attDirName,
          })
          msgEntries.push([attDirName, attDirEntry])
          const attEntries: [string, IndexEntry][] = []
          for (const att of attachments) {
            const attEntry = new IndexEntry({
              id: att.filename,
              name: att.filename,
              resourceType: 'email/attachment',
              vfsName: att.filename,
              size: att.size,
            })
            attEntries.push([att.filename, attEntry])
          }
          const attDirVKey = `${virtualKey}/${dateStr}/${attDirName}`
          await index.setDir(attDirVKey, attEntries)
        }
      }
      const dateVKey = `${virtualKey}/${dateStr}`
      await index.setDir(dateVKey, msgEntries)
    }
    await index.setDir(virtualKey, dateEntries)
    return dateEntries.map(([name]) => `${prefix}/${key}/${name}`)
  }

  if (depth === 2 || depth === 3) {
    if (index === undefined) throw enoent(path.original)
    let cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    const folderPath = prefix !== '' ? `${prefix}/${parts[0] ?? ''}` : `/${parts[0] ?? ''}`
    const folderSpec = new PathSpecCtor({ original: folderPath, directory: folderPath, prefix })
    await readdir(accessor, folderSpec, index)
    cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    throw enoent(path.original)
  }

  throw enoent(path.original)
}

File: python/mirage/core/email/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import re
from email.utils import parsedate_to_datetime

from mirage.accessor.email import EmailAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.email._client import fetch_headers, list_message_uids
from mirage.core.email.folders import list_folders
from mirage.types import PathSpec
from mirage.utils.errors import enoent

TITLE_MAX = 80
UNSAFE = re.compile(r"[^\w\s\-.]")
MULTI_UNDERSCORE = re.compile(r"_+")


def _sanitize(text: str) -> str:
    if not text.strip():
        return "No_Subject"
    cleaned = UNSAFE.sub("_", text).replace(" ", "_")
    cleaned = MULTI_UNDERSCORE.sub("_", cleaned).strip("_")
    if len(cleaned) > TITLE_MAX:
        cleaned = cleaned[:TITLE_MAX - 3] + "..."
    return cleaned


def _msg_filename(subject: str, uid: str) -> str:
    return f"{_sanitize(subject)}__{uid}.email.json"


def _date_from_header(date_str: str) -> str:
    try:
        dt = parsedate_to_datetime(date_str)
        return dt.strftime("%Y-%m-%d")
    except Exception:
        return "1970-01-01"


async def readdir(
    accessor: EmailAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.directory if path.pattern else path.original
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    virtual_key = prefix + "/" + key if key else prefix or "/"
    parts = key.split("/") if key else []
    depth = len(parts)

    if depth == 0:
        if index is not None:
            cached = await index.list_dir(virtual_key)
            if cached.entries is not None:
                return cached.entries
        folders = await list_folders(accessor)
        entries = []
        for folder_name in folders:
            entry = IndexEntry(
                id=folder_name,
                name=folder_name,
                resource_type="email/folder",
                vfs_name=folder_name,
            )
            entries.append((folder_name, entry))
        if index is not None:
            await index.set_dir(virtual_key, entries)
        return [f"{prefix}/{name}" for name, _ in entries]

    if depth == 1:
        folder_name = parts[0]
        if index is not None:
            cached = await index.list_dir(virtual_key)
            if cached.entries is not None:
                return cached.entries
        if index is None:
            raise enoent(virtual)
        max_msgs = accessor.config.max_messages
        uids = await list_message_uids(accessor,
                                       folder_name,
                                       max_results=max_msgs)
        headers_list = await fetch_headers(accessor, folder_name, uids)
        date_groups: dict[str, list[dict]] = {}
        for hdr in headers_list:
            date_str = _date_from_header(hdr.get("date", ""))
            date_groups.setdefault(date_str, []).append(hdr)
        date_entries: list[tuple[str, IndexEntry]] = []
        for date_str in sorted(date_groups.keys(), reverse=True):
            date_entry = IndexEntry(
                id=date_str,
                name=date_str,
                resource_type="email/date",
                vfs_name=date_str,
            )
            date_entries.append((date_str, date_entry))
            msg_entries: list[tuple[str, IndexEntry]] = []
            for hdr in date_groups[date_str]:
                uid = hdr["uid"]
                subject = hdr.get("subject", "") or "No Subject"
                filename = _msg_filename(subject, uid)
                msg_entry = IndexEntry(
                    id=uid,
                    name=subject,
                    resource_type="email/message",
                    vfs_name=filename,
                )
                msg_entries.append((filename, msg_entry))
                attachments = hdr.get("attachments", [])
                if attachments:
                    att_dir_name = filename.replace(".email.json", "")
                    att_dir_entry = IndexEntry(
                        id=uid,
                        name=att_dir_name,
                        resource_type="email/attachment_dir",
                        vfs_name=att_dir_name,
                    )
                    msg_entries.append((att_dir_name, att_dir_entry))
                    att_entries: list[tuple[str, IndexEntry]] = []
                    for att in attachments:
                        att_entry = IndexEntry(
                            id=att["filename"],
                            name=att["filename"],
                            resource_type="email/attachment",
                            vfs_name=att["filename"],
                            size=att.get("size"),
                        )
                        att_entries.append((att["filename"], att_entry))
                    att_dir_vkey = (virtual_key + "/" + date_str + "/" +
                                    att_dir_name)
                    await index.set_dir(att_dir_vkey, att_entries)
            date_vkey = virtual_key + "/" + date_str
            await index.set_dir(date_vkey, msg_entries)
        await index.set_dir(virtual_key, date_entries)
        return [f"{prefix}/{key}/{name}" for name, _ in date_entries]

    if depth == 2:
        if index is None:
            raise enoent(virtual)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        folder_vkey = PathSpec(
            original=prefix + "/" + parts[0],
            directory=prefix + "/" + parts[0],
            prefix=prefix,
        )
        await readdir(accessor, folder_vkey, index)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        raise enoent(virtual)

    if depth == 3:
        if index is None:
            raise enoent(virtual)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        folder_vkey = PathSpec(
            original=prefix + "/" + parts[0],
            directory=prefix + "/" + parts[0],
            prefix=prefix,
        )
        await readdir(accessor, folder_vkey, index)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        raise enoent(virtual)

    raise enoent(virtual)
